File: internal/youtube/paging.go
package youtube

import "strings"

// TranscriptPage is a bounded window over a transcript. Agents with limited
// context ask for maxChars at a time and follow nextCursor until hasMore is false.
type TranscriptPage struct {
	Transcript
	Cursor        int  `json:"cursor"`
	NextCursor    int  `json:"nextCursor,omitempty"`
	HasMore       bool `json:"hasMore"`
	TotalSegments int  `json:"totalSegments"`
	TotalChars    int  `json:"totalChars"`
	PageChars     int  `json:"pageChars"`
}

// PageTranscript slices tr starting at segment index cursor, stopping once the
// page would exceed maxChars. A segment is never split, so a single oversized
// segment is still returned whole.
func PageTranscript(tr *Transcript, cursor, maxChars int) *TranscriptPage {
	if cursor < 0 {
		cursor = 0
	}
	if cursor > len(tr.Segments) {
		cursor = len(tr.Segments)
	}
	totalChars := 0
	for _, s := range tr.Segments {
		totalChars += len([]rune(s.Text))
	}

	end := len(tr.Segments)
	pageChars := 0
	if maxChars > 0 {
		end = cursor
		for end < len(tr.Segments) {
			segChars := len([]rune(tr.Segments[end].Text))
			if pageChars > 0 && pageChars+segChars > maxChars {
				break
			}
			pageChars += segChars
			end++
		}
	} else {
		for _, s := range tr.Segments[cursor:] {
			pageChars += len([]rune(s.Text))
		}
	}

	segments := tr.Segments[cursor:end]
	lines := make([]string, len(segments))
	var text strings.Builder
	var duration float64
	for i, s := range segments {
		lines[i] = "[" + s.Timestamp + "] " + s.Text
		if i > 0 {
			text.WriteByte(' ')
		}
		text.WriteString(s.Text)
		if s.End > duration {
			duration = s.End
		}
	}

	page := &TranscriptPage{
		Transcript:    *tr,
		Cursor:        cursor,
		HasMore:       end < len(tr.Segments),
		TotalSegments: len(tr.Segments),
		TotalChars:    totalChars,
		PageChars:     pageChars,
	}
	page.Segments = segments
	page.Lines = lines
	page.Text = text.String()
	page.SegmentCount = len(segments)
	page.DurationSeconds = duration
	if page.HasMore {
		page.NextCursor = end
	}
	return page
}

File: src/engine/paging.ts
/**
 * Bounded windows over a transcript, for agents with limited context.
 * Ported from internal/youtube/paging.go.
 */
import type { Transcript, TranscriptPage } from "./types.js";

export type { TranscriptPage } from "./types.js";

function runeLength(text: string): number {
  return [...text].length;
}

/**
 * Slices `tr` starting at segment index `cursor`, stopping once the page would
 * exceed `maxChars`. A segment is never split, so a single oversized segment is
 * still returned whole. `maxChars <= 0` returns everything from the cursor on.
 */
export function pageTranscript(tr: Transcript, cursor: number, maxChars: number): TranscriptPage {
  const all = tr.segments;
  let start = cursor;
  if (start < 0) {
    start = 0;
  }
  if (start > all.length) {
    start = all.length;
  }

  let totalChars = 0;
  for (const s of all) {
    totalChars += runeLength(s.text);
  }

  let end = all.length;
  let pageChars = 0;
  if (maxChars > 0) {
    end = start;
    while (end < all.length) {
      const segChars = runeLength(all[end]!.text);
      if (pageChars > 0 && pageChars + segChars > maxChars) {
        break;
      }
      pageChars += segChars;
      end++;
    }
  } else {
    for (const s of all.slice(start)) {
      pageChars += runeLength(s.text);
    }
  }

  const segments = all.slice(start, end);
  const lines: string[] = [];
  const textParts: string[] = [];
  let duration = 0;
  for (const s of segments) {
    lines.push("[" + s.timestamp + "] " + s.text);
    textParts.push(s.text);
    if (s.end > duration) {
      duration = s.end;
    }
  }

  const hasMore = end < all.length;
  return {
    ...tr,
    segmentCount: segments.length,
    durationSeconds: duration,
    segments,
    lines,
    text: textParts.join(" "),
    cursor: start,
    nextCursor: hasMore ? end : undefined,
    hasMore,
    totalSegments: all.length,
    totalChars,
    pageChars,
  };
}
